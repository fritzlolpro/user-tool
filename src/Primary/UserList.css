.list-wrapper {
  display: flex;
  flex-wrap: wrap;
  align-content: flex-start;
  margin-bottom: 16px;
}
.user-card {
  margin-left: 16px !important;
}
.standart-card {
  margin-bottom: 24px !important;
  padding: 16px !important;
  width: 450px !important;
  height: 550px !important;
  transition: .7 !important;
}
