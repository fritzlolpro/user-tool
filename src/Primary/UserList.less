.list-wrapper {
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
    margin-bottom: 16px;
}

.standart-card() {
    margin-bottom: 24px;
    padding: 16px;
    width: 450px;
    height: 550px;
    transition: .7;
}

.user-card() {
    margin-left: 16px;
}

.user-card {
    .user-card() !important
}

.standart-card {
    .standart-card() !important;
}