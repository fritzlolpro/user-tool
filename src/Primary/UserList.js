import React, {Component} from 'react';
import UserInfoEditor from '../Forms/UserInfoEditor.js'
import UserInfo from '../Output/UserInfo.js'
import ErrorText from '../Output/Errors.js'
import Paper from 'material-ui/Paper';
import Snackbar from 'material-ui/Snackbar';
import './UserList.css';

export default class UserList extends Component {
    constructor(props) {
        super(props)
        this.state = {
            editing: null,
            update: null,
            users: [],
            error: false,
            notificationOpen: false,
            notificationMessage: ''
        }

    }

    callUsersApi = (options, path) => this
        .props
        .apiCall(options, path)

    setUserList = (users) => {
        this.setState({users})
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.newUser !== this.props.newUser) {
            this.getUsersFromServer()
        }

    }

    handleNotificationsClose = () => {
        this.setState({notificationOpen: false});
    }

    handleErrors = (response) => {
        if (!response.ok) {
            this.setState({error: response.statusText})
        }

        return response;
    }

    onDismissError = () => {
        this.setState({error: false})
    }

    getUsersFromServer() {
        this
            .callUsersApi()
            .then(response => this.handleErrors(response))
            .then(blob => blob.json())
            .then(users => this.setUserList(users))
            .catch(error => this.setState({
                error: error.toString()
            }))
    }

    componentDidMount() {
        this.getUsersFromServer()
    }

    renderUserOrUserEditor(user) {
        if (this.state.editing === user.id) {
            // render editor for respectin' user use dull key for initial render
            return (
                <Paper key={user.id || 1} className='standart-card user-card' zDepth={3}>
                    <UserInfoEditor
                        user={user}
                        update={this.state.update}
                        onChange={this.onChangeField}
                        onSubmit={this.onSubmitField}
                        onCancel={this.onCancelChangeField}/>
                </Paper>
            )
        } else {
            // render user component
            return (
                <Paper key={user.id || 1} className='standart-card user-card'>
                    <UserInfo user={user} onEdit={this.onEdit} onDelete={this.onDelete}/>
                </Paper>
            )
        }

    }

    onEdit = (id) => {
        this.setState({editing: id})
    }

    onDelete = (id) => {

        const options = {
            method: 'delete'
        }

        this
            .callUsersApi(options, `/${id}`)
            .then(this.setUserList(this.generateShrinkedUsersList(id)), this.setState({notificationMessage: 'The user has been deleted'}), this.setState(prevState => ({notificationOpen: true})))
    }

    onChangeField = (e) => {
        const target = e.target
        const value = target.value
        const name = target.name
        const {update} = this.state || []
        this.setState({
            update: {
                ...update,
                [name]: value
            }

        })
    }

    generateShrinkedUsersList = (deletedUserId) => {
        // if some user was deleted, delete him from list
        const {users} = this.state
        const isNotDeletedUser = user => user.id !== deletedUserId
        return users.filter(isNotDeletedUser)
    }

    genrateUpdatedUserList = (newUserData) => {
        // if some user info was changed we can figure what and then update user list
        const {users} = this.state
        return users.map(x => {
            if (x.id === newUserData.id) {
                return (x = {
                    ...newUserData
                })
            } else {
                return x
            }

        })
    }

    onSubmitField = (user) => {
        const {update} = this.state
        const data = {
            id: user.id,
            name: {
                first: update && update.first
                    ? update.first
                    : user.name.first,
                last: update && update.last
                    ? update.last
                    : user.name.last
            },
            picture: update && update.picture
                ? update.picture
                : user.picture
        }

        const options = {
            method: 'put',
            body: JSON.stringify(data)
        }

        this
            .callUsersApi(options, `/${user.id}`)
            .then(response => this.handleErrors(response))
            .then(this.onCancelChangeField(), this.setUserList(this.genrateUpdatedUserList(data)), this.setState({notificationMessage: 'The user has been edited'}), this.setState(prevState => ({notificationOpen: true})))
            .catch(error => this.setState({
                error: error.toString()
            }))
    }

    onCancelChangeField = (e) => {
        this.setState({editing: null, update: null})
    }

    render() {
        const {users, error, notificationOpen, notificationMessage} = this.state;

        const list = users || []
        return <div className='list-wrapper'>
            {error
                ? <ErrorText onClose={this.onDismissError} error={error}>Problems while recieving user list
                        <br/>
                        Server says: {error}</ErrorText>
                : list.map(user => this.renderUserOrUserEditor(user))}

            <Snackbar
                open={notificationOpen}
                message={notificationMessage}
                autoHideDuration={4000}
                onRequestClose={this.handleNotificationsClose}/>
        </div>
    }
}
