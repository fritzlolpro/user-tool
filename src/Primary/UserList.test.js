import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import UserList from './UserList';

import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';

describe('UserList', () => {

    const mockCall = () => {
        return new Promise((resolve, reject) => {
            resolve(jest.fn())
        })
    }
    const props = {
        users: [
            {
                id: 123,
                name: {
                    first: 'TestFirst',
                    last: 'TestLast'
                },
                picture: 'https://randomuser.me/api/portraits/men/83.jpg'
            }, {
                id: 56,
                name: {
                    first: 'TestFirst2',
                    last: 'TestLast2'
                },
                picture: 'https://randomuser.me/api/portraits/men/33.jpg'
            }
        ],
        apiCall: () => mockCall()

    }

    it('should be defined', () => {
        expect(UserList).toBeDefined();
    });

    it('renders without crashing'),
    () => {
        const div = document.createElement('div');
        ReactDOM.render(
            <MuiThemeProvider>
            <UserList {...props}/>
        </MuiThemeProvider>, div);
    }

    test('has a valid snapshot', () => {
        const component = renderer.create(
            <MuiThemeProvider>
                <UserList {...props}/>
            </MuiThemeProvider>
        )
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    })
})