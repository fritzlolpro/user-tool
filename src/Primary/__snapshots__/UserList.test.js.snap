// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList has a valid snapshot 1`] = `
<div
  className="list-wrapper"
>
  <div
    open={false}
    style={
      Object {
        "bottom": 0,
        "display": "flex",
        "left": "50%",
        "muiPrepared": true,
        "position": "fixed",
        "transform": "translate(-50%, 48px)",
        "transition": "transform 400ms cubic-bezier(0.23, 1, 0.32, 1) 0ms, visibility 400ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
        "visibility": "hidden",
        "zIndex": 2900,
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.87)",
          "borderRadius": 2,
          "flexGrow": 0,
          "fontFamily": "Roboto, sans-serif",
          "height": 48,
          "lineHeight": "48px",
          "maxWidth": 568,
          "minWidth": 288,
          "muiPrepared": true,
          "padding": "0 24px",
          "width": "auto",
        }
      }
      width={3}
    >
      <div
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 14,
            "muiPrepared": true,
            "opacity": 0,
            "transition": "opacity 400ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          }
        }
      >
        <span>
          
        </span>
      </div>
    </div>
  </div>
</div>
`;
